File: php/src/Inbox.php
<?php

declare(strict_types=1);

namespace Sentroy\ClientSdk;

class Inbox
{
  private HttpClient $http;

  public function __construct(HttpClient $http)
  {
    $this->http = $http;
  }

  /**
   * List messages in a mailbox folder.
   *
   * @param array{mailbox?: string, folder?: string, page?: int, limit?: int, unreadOnly?: bool} $params
   */
  public function list(array $params = []): array
  {
    $query = [];
    if (isset($params["mailbox"])) $query["mailbox"] = $params["mailbox"];
    if (isset($params["folder"])) $query["folder"] = $params["folder"];
    if (isset($params["page"])) $query["page"] = $params["page"];
    if (isset($params["limit"])) $query["limit"] = $params["limit"];
    if (!empty($params["unreadOnly"])) $query["unreadOnly"] = "true";

    return $this->http->get("/inbox", $query);
  }

  /**
   * Get a single message detail.
   *
   * @param array{mailbox?: string, folder?: string} $options
   */
  public function get(int $uid, array $options = []): array
  {
    $query = [];
    if (isset($options["mailbox"])) $query["mailbox"] = $options["mailbox"];
    if (isset($options["folder"])) $query["folder"] = $options["folder"];

    return $this->http->get("/inbox/{$uid}", $query);
  }

  /** List IMAP folders (mailboxes) for a given email account */
  public function listFolders(?string $mailbox = null): array
  {
    $query = [];
    if ($mailbox !== null) $query["mailbox"] = $mailbox;

    return $this->http->get("/inbox/mailboxes", $query);
  }

  /** Get thread messages by subject */
  public function getThread(string $subject, ?string $mailbox = null): array
  {
    $query = ["subject" => $subject];
    if ($mailbox !== null) $query["mailbox"] = $mailbox;

    return $this->http->get("/inbox/thread", $query);
  }

  /**
   * Mark a message as read.
   *
   * @param array{mailbox?: string, folder?: string} $options
   */
  public function markAsRead(int $uid, array $options = []): void
  {
    $this->http->post("/inbox/{$uid}/read", [
      "mailbox" => $options["mailbox"] ?? null,
      "folder" => $options["folder"] ?? null,
    ]);
  }

  /**
   * Mark a message as unread.
   *
   * @param array{mailbox?: string, folder?: string} $options
   */
  public function markAsUnread(int $uid, array $options = []): void
  {
    $query = [];
    if (isset($options["mailbox"])) $query["mailbox"] = $options["mailbox"];
    if (isset($options["folder"])) $query["folder"] = $options["folder"];

    $this->http->delete("/inbox/{$uid}/read", $query);
  }

  /**
   * Move a message to another folder.
   *
   * @param array{from?: string, mailbox?: string} $options
   */
  public function move(int $uid, string $to, array $options = []): void
  {
    $this->http->post("/inbox/{$uid}/move", [
      "to" => $to,
      "from" => $options["from"] ?? null,
      "mailbox" => $options["mailbox"] ?? null,
    ]);
  }

  /**
   * Delete a message.
   *
   * @param array{mailbox?: string, folder?: string} $options
   */
  public function delete(int $uid, array $options = []): void
  {
    $query = [];
    if (isset($options["mailbox"])) $query["mailbox"] = $options["mailbox"];
    if (isset($options["folder"])) $query["folder"] = $options["folder"];

    $this->http->delete("/inbox/{$uid}", $query);
  }
}

File: go/inbox.go
package sentroy

import (
	"fmt"
	"net/http"
)

// InboxService provides access to inbox-related API endpoints.
type InboxService struct {
	h *httpClient
}

func newInboxService(h *httpClient) *InboxService {
	return &InboxService{h: h}
}

// InboxGetOptions are optional parameters for single-message operations.
type InboxGetOptions struct {
	Mailbox string
	Folder  string
}

// InboxMoveOptions are optional parameters for the Move operation.
type InboxMoveOptions struct {
	From    string
	Mailbox string
}

// List returns messages in a mailbox folder.
func (s *InboxService) List(params *InboxListParams) ([]MessageSummary, error) {
	q := make(map[string]string)
	if params != nil {
		if params.Mailbox != "" {
			q["mailbox"] = params.Mailbox
		}
		if params.Folder != "" {
			q["folder"] = params.Folder
		}
		if params.Page > 0 {
			q["page"] = fmt.Sprintf("%d", params.Page)
		}
		if params.Limit > 0 {
			q["limit"] = fmt.Sprintf("%d", params.Limit)
		}
		if params.UnreadOnly {
			q["unreadOnly"] = "true"
		}
	}
	return doRequest[[]MessageSummary](s.h, http.MethodGet, "/inbox", q, nil)
}

// Get returns the full detail of a single message.
func (s *InboxService) Get(uid int, opts *InboxGetOptions) (*MessageDetail, error) {
	q := make(map[string]string)
	if opts != nil {
		if opts.Mailbox != "" {
			q["mailbox"] = opts.Mailbox
		}
		if opts.Folder != "" {
			q["folder"] = opts.Folder
		}
	}
	d, err := doRequest[MessageDetail](s.h, http.MethodGet, fmt.Sprintf("/inbox/%d", uid), q, nil)
	if err != nil {
		return nil, err
	}
	return &d, nil
}

// ListFolders returns the IMAP folders (mailboxes) for a given email account.
func (s *InboxService) ListFolders(mailbox string) ([]Mailbox, error) {
	q := make(map[string]string)
	if mailbox != "" {
		q["mailbox"] = mailbox
	}
	return doRequest[[]Mailbox](s.h, http.MethodGet, "/inbox/mailboxes", q, nil)
}

// GetThread returns all messages in a thread identified by subject.
func (s *InboxService) GetThread(subject string, mailbox string) ([]MessageDetail, error) {
	q := map[string]string{"subject": subject}
	if mailbox != "" {
		q["mailbox"] = mailbox
	}
	return doRequest[[]MessageDetail](s.h, http.MethodGet, "/inbox/thread", q, nil)
}

// MarkAsRead marks a message as read (sets the \Seen flag).
func (s *InboxService) MarkAsRead(uid int, opts *InboxGetOptions) error {
	body := make(map[string]interface{})
	if opts != nil {
		if opts.Mailbox != "" {
			body["mailbox"] = opts.Mailbox
		}
		if opts.Folder != "" {
			body["folder"] = opts.Folder
		}
	}
	_, err := doRequest[interface{}](s.h, http.MethodPost, fmt.Sprintf("/inbox/%d/read", uid), nil, body)
	return err
}

// MarkAsUnread marks a message as unread (removes the \Seen flag).
func (s *InboxService) MarkAsUnread(uid int, opts *InboxGetOptions) error {
	q := make(map[string]string)
	if opts != nil {
		if opts.Mailbox != "" {
			q["mailbox"] = opts.Mailbox
		}
		if opts.Folder != "" {
			q["folder"] = opts.Folder
		}
	}
	_, err := doRequest[interface{}](s.h, http.MethodDelete, fmt.Sprintf("/inbox/%d/read", uid), q, nil)
	return err
}

// Move moves a message to another folder.
func (s *InboxService) Move(uid int, to string, opts *InboxMoveOptions) error {
	body := map[string]interface{}{"to": to}
	if opts != nil {
		if opts.From != "" {
			body["from"] = opts.From
		}
		if opts.Mailbox != "" {
			body["mailbox"] = opts.Mailbox
		}
	}
	_, err := doRequest[interface{}](s.h, http.MethodPost, fmt.Sprintf("/inbox/%d/move", uid), nil, body)
	return err
}

// Delete deletes a message.
func (s *InboxService) Delete(uid int, opts *InboxGetOptions) error {
	q := make(map[string]string)
	if opts != nil {
		if opts.Mailbox != "" {
			q["mailbox"] = opts.Mailbox
		}
		if opts.Folder != "" {
			q["folder"] = opts.Folder
		}
	}
	_, err := doRequest[interface{}](s.h, http.MethodDelete, fmt.Sprintf("/inbox/%d", uid), q, nil)
	return err
}
